### DISCUSSION TITLE
Imperative vs Declarative mapping style in Domain Driven Design project

### DISCUSSION THREAD

**Discussion by toinbis** [sqlalchemy/sqlalchemy#9180] [Usage Questions] [answered]
Hey there! Congrats to everyone on v2 launch!

I want to stick to DDD best practices and define domain entities/value objects in python (data)classes, which have nothing to do with persistance. Can't decide should I use imperative or declarative mapping style for SQLAlchemy ORM.

https://github.com/qu3vipon/python-ddd/blob/main/src/shared_kernel/infra/database/orm.py , https://github.com/qu3vipon/python-ddd/blob/dc4972673c95f2b794a7c9064db6e1e569fd2f4a/src/shared_kernel/infra/database/repository.py#L13 - this is a repo from which I intend to borrow the implementation details if I choose the imperative style. You see the tables explicitly set, you map the domain entities to table using registry mapper, and then can use domain entities with `session.add(instance)`. All clear and nice. For me the only real minus is that in a mapper properties the `table.c.column` the `column` part does not have type hints in editor, but I can live with that.

**My question is - what would I do if I wanted to use declarative mapping style?**

**Write my own mapper from scratch?** Tried it! With `persistance_entity = MyOrmTableSublassedFromBase(**dict(domain_entity))`, where domain_entity is pydantic model with `orm_mode = True`. When I tried to add `persistance_entity` to session, the error pops up:

```
sqlalchemy/orm/attributes.py", line 1739, in fire_append_event
    self.sethasparent(instance_state(value), state, True)
                      ^^^^^^^^^^^^^^^^^^^^^
AttributeError: <...> object has no attribute '_sa_instance_state'
```
And chances are going down this path I loose most of the ORM benefits recreating the registry myself from scratch (maybe not, not sure).

**Or maybe create "secondary" mapping between entity class and Orm Base subclass? Maybe this would be a recommended approach for declarative mapping style?**

So ultimately my question comes down to:
**a) Is there a preffered/recommended mapping style for scenario when domain entities are not related with SQLA anyhow and the mapping needs to be done?
b) If imperative is prefered - all clear, I have an example to follow. If they are ~equal, what's the recommended scenario to map domain entities to declarative base class - use the registry or write my own custom mapper?**

**Comment by zzzeek** [comment, marked as answer]
> a) Is there a preffered/recommended mapping style for scenario when domain entities are not related with SQLA anyhow and the mapping needs to be done?

you would use imperative mapping for that, however I don't recommend this pattern overall.  SQLAlchemy mappings modify the behavior of the class and the object so trying to keep that all "separate" does not really accomplish anything IMO.  I was hoping to remove imperative mapping from 2.0 entirely **EDIT: PLEASE NOTE IT IS NOT BEING REMOVED AND IT REMAINS FULLY SUPPORTED AND DOCUMENTED** but a lot of people still like it a lot, so you are supported with this pattern.


> b) If imperative is prefered - all clear, I have an example to follow. If they are ~equal, what's the recommended scenario to map domain entities to declarative base class - use the registry or write my own custom mapper?

I'm not following what the "write my own custom mapper" case means without examples, but it certainly sounds like something formidable and not within what we'd want our users to be doing.

**Reply by mr-mapache** [reply]
I don't agree. When you have full control over you database schemas, declarative is the way to go, but let's say you have to develop an app over an already existing database that you don't control and your domain layer (python dataclasses) don't have anything to do with your data, there is where imperative style shines, otherwise you end up making hundreds of dtos and factories for the declarative schemas.

Take this as an example, you have a User aggregate root with a List[Event] member, and a Publisher member to publish those events, that User has to be registered into you system using a phone number and publishing the event into the message queue, but your database have users and emails in the same table and phone numbers in another table, how are you supposed to solve that using the "new and shining declarative style"? You would be creating hundreds of dtos, or using raw sql if it weren't for the imperative style that makes this kind of stuff easy. 

I think is a big mistake to drop support for imperative style, since it is what makes sqlalchemy different from other orms. Otherwise we all would be using django.

**Reply by CaselIT** [reply]
please note that once the classes are mapped they are exactly the same independently of how you have mapped them. So if you can solve something using imperative, you can do the exact same thing using declarative.

**Reply by mr-mapache** [reply]
> please note that once the classes are mapped they are exactly the same independently of how you have mapped them. So if you can solve something using imperative, you can do the exact same thing using declarative.

I was doing something like for some time that but the came back to the imperative mapping style because what I was doing was basically code smell.

Let's say I have a users table in a database with the following style:

```python
users = Table(
    'users',
    metadata,
    Column('id', Integer, primary_key=True), # surrogate key, don't have anything to do with the domain id.
    Column('user_id', UUID, unique=True, nullable=False),
    Column('user_username', String, unique=True),
)
```

This would be the declarative style:

```python
class Schema(DeclarativeBase):
    surrogate_key: Mapped[int] = mapped_column('id', primary_key=True) #have to rename this key from database.

class User(Schema):
    __tablename__ = 'users'
    id: Mapped[UUID] = mapped_column('user_id', unique=True, nullable=False)
    username: Mapped[str] = mapped_column('user_username', unique=True, nullable=False)
```

As you see, in the first one, it's very clear I'm talking about a table, that shouldn't be initialized, and has to be used with the insert, select, update, delete style. But in the second one example, is very easy for another person to confuse the User "table" with the User aggregate, and cannot import those two without an alias as their names colide.

An easy solution for having PEP support was to mimic imperative style with declarative.

```python
class users(Schema):
    __tablename__ = 'users'
    id: Mapped[int] = mapped_column('id', primary_key=True)
    user_id: Mapped[UUID] = mapped_column('user_id', unique=True)
    user_username: Mapped[str] = mapped_column('user_username', unique=True, nullable=False)
```

But I don't really think that it's okay to do that, declarative mapping is not supposed to be used in that way, creates very confusing code. Shouldn't be to hard to create something like a factory of classes returning something like the above with PEP support. 

It's sad, since sqlalchemy imperative style it's some of the very few orm's that solves the domain-orm impedance mismatch problem.

**Reply by mr-mapache** [reply]
> please note that once the classes are mapped they are exactly the same independently of how you have mapped them. So if you can solve something using imperative, you can do the exact same thing using declarative.

Call me crazy, but you have a bunch of people using raw sql, creating custom mappers or doing nasty things like I was, look at this [guy](https://www.reddit.com/r/Python/comments/rt3apo/developing_a_python_database_orm_with_a_different/) I really think that this is an issue.

**Reply by CaselIT** [reply]
I don't understand what's the point you are trying to make with your user examples. If you don't want an orm class but you want to use a table, use a table. That's not what imperative style is.

**Comment by toinbis** [comment]
@zzzeek - prompt and detailed answer, as always. Many thanks for that! Things are much clearer now.

"My own custom mapper" -> https://github.com/NEONKID/fastapi-ddd-example/blob/b6652bf5d0a61f693ce0ffb947e1e8cb331bd35c/modules/author/infrastructure/persistence/mapper.py#L11, simple Data Mapper pattern implementation, maping domain to persistance entity and back.

I guess for my use case I'll have to make a choice between imperative ORM mapping style or SQLAlchemy core. Won't be an easy pick, but thanks again for providing both options! :)

**Reply by CaselIT** [reply]
> "My own custom mapper" -> https://github.com/NEONKID/fastapi-ddd-example/blob/b6652bf5d0a61f693ce0ffb947e1e8cb331bd35c/modules/author/infrastructure/persistence/mapper.py#L11, simple Data Mapper pattern implementation, maping domain to persistance entity and back.

Not sure I follow that example. `Author` is also mapped imperatively by `start_mapper`, so you have both `Author` and ` AuthorEntity` that are mapped class, that I can't see how it make any sense.

You mentioned pydantic. It supports something similar (in only one direction persistence -> domain), where only one of the classes is mapped, not both. https://docs.pydantic.dev/usage/models/#orm-mode-aka-arbitrary-class-instances

**Reply by zzzeek** [reply]
> @zzzeek - prompt and detailed answer, as always. Many thanks for that! Things are much clearer now.
> 
> "My own custom mapper" -> https://github.com/NEONKID/fastapi-ddd-example/blob/b6652bf5d0a61f693ce0ffb947e1e8cb331bd35c/modules/author/infrastructure/persistence/mapper.py#L11, simple Data Mapper pattern implementation, maping domain to persistance entity and back.


oh ok, yeah pydantic has that sort of thing as well.   the code you have can work but adds a palpable performance tax to do those conversions each time.  pydantic might be better since it's largely written in Rust (or at least will be, not sure).

**Reply by toinbis** [reply]
Thanks for great discussion @CaselIT @zzzeek !

- As for the example, https://github.com/NEONKID/fastapi-ddd-example#sqlalchemy-classical-mapping - here the owner of repo explains that both (all!:) ways of mapping are present just as a learning example. `AuthorMapper` class is defined but used nowhere elso in the repo, imperative mapping initiating function `start_mapper` is used all over the place. 
- I'm literally in the same position as the ^ repo creator - having a repository uses both sqlalchemy orm/sqlalchemy core to store/retrieve data (and compare to make sure it's a match!:). Now it'll have two orm styles and core, haha :) It's a very painful decision to make. Hope it wont take longer than one more month to make a decision what to use:)
- Domain model IMO should be as pure as it can get - would not be willing to bet on specific pydantic functionality to be used.
- Today I googled out that `Architecture Patterns with Python` book ( https://www.amazon.com/Architecture-Patterns-Python-Domain-Driven-Microservices/dp/1492052205 ) in their code examples  https://github.com/cosmicpython/code/blob/master/src/allocation/adapters/orm.py also sort of recommend to use imperative mapping style. I suppose for DDD purists still willing to get the benefits of ORM imperative mapping is still a great compromise.

Thanks again for discussion and your work on SQLAlchemy!

**Reply by CaselIT** [reply]
> Today I googled out that Architecture Patterns with Python book ( https://www.amazon.com/Architecture-Patterns-Python-Domain-Driven-Microservices/dp/1492052205 ) in their code examples https://github.com/cosmicpython/code/blob/master/src/allocation/adapters/orm.py also sort of recommend to use imperative mapping style. I suppose for DDD purists still willing to get the benefits of ORM imperative mapping is still a great compromise.

please note that the example is mapping domain models with sqlalchemy imperative mapping. This does the same thing as the declarative mapping, meaning that the domain models **are** mapped sqlalchemy classes.

If you are ok with this, then you should be ok also with declarative mapping, since the only difference between the two is syntactical, at runtime they are the same exact thing

If you want to keep things separated, you necessarily need two classes, and need to translate between one instance and the other.
I mean that if you use the domain models to query the db with the sqlalchemy, like in the example above https://github.com/cosmicpython/code/blob/69a88f8e05d549cc4cf01a91cd33b0fc4d87014d/src/allocation/adapters/repository.py#L48-L49, then the domain model and the orm model *are* the same thing, so from my point of view you may as well use the declarative style, that in v2 is basically similar to creating a dataclass

**Reply by zzzeek** [reply]
I believe DDD seeks a certain flavor of "purity" where the model classes are declared such that any indication that they are linked to a database ORM is not explicitly stated.

This certainly was something I was going for back when I first wrote SQLAlchemy.   However, around version 0.5 or so, which is pretty much the first time I actually had a paid job where I got to use SQLAlchemy for real, I immediately wrote Declarative as the impracticality of using mapper() became apparent.  At that time, there was a library called Elixir which provided the previous form of "Declarative", which itself had a very difficult task in providing this functionality on the outside, so I created something similar that could be more integrated with SQLAlchemy.

As mentioned before though, there's a lot of people that still like to use mapper() (now called `map_imperatively()`), so even though it seems pretty impractical to me, other people don't think so.

**Reply by lelit** [reply]
I'm one of those :wink:

In a couple of cases, I used different sets of classes, implementing different functionalities that for various reasons I preferred to _isolate_ in completely unrelated hierarchies, while having the underlying tables in a common module. I still keep that habit for most of my non-trivial apps.

**Reply by horvatha** [reply]
@CaselIT  Than how can I swap relational database to a NoSQL one later. With the Pattern used in Harry's book it is easy. I just write the Repository and Unit of Work pattern for NoSQL without touching the domain model and the need of importing SQLA is gone. Also there is an [example in the Appendix](https://www.cosmicpython.com/book/appendix_csvs.html), where they swap the database to CSV files, and the Flask app to a CLI, and why would they install SQLAlchemy at all for this use case. They don't touch the domain model at all (again just new Repo and UoW classes).

(Also using [FakeRepository](https://github.com/cosmicpython/code/blob/chapter_06_uow/tests/unit/test_services.py#L6) in tests will be more complicated, I guess.)

**Reply by CaselIT** [reply]
> how can I swap relational database to a NoSQL one later

In usual applications such changes are not frequent and I doubt how the mapping is done is what causes most of the work 

In any both ways of mapping things are available, so use what suits your use case most

**Reply by qu3vipon** [reply]
First of all, thank you for referring to my code example. 

I also prefer the imperative mapping style. There are two main reasons.

First, the model can be separated by bounded context. 

If you look at [the code example](https://github.com/qu3vipon/python-ddd/blob/main/src/shared_kernel/infra/database/orm.py#L38), you can see that one `room table` is mapped to two entities based on different bounded contexts. 

This is because even if you share the same data in different bounded contexts, it can be used in different contexts. (In general, it is better to modify the data only in one context, and only do read operations in the other context because it does not produce side effects.)

And you can use it when you apply CQRS patterns.

Second, the Test runs very quickly. 

The database I/O is often used because of the difficulty of mocking when using declarative mapping, but if the project grows and the number of test cases increases, the speed at which the test runs increases significantly. 
On the other hand, imperative mapping is much more useful because it allows you to process tests on memory quickly without DB I/O.

**Reply by CaselIT** [reply]
> Second, Test runs very quickly.

Not sure I follow this one. If you want to test db io but you are not using the db then the test is useless.

I'm probably not following what you mean

**Reply by qu3vipon** [reply]
The term "test" was too comprehensive. When testing the functionality of the web server API, sometimes you don't have to make DB I/O, which means that you can handle everything on the memory.

**Reply by CaselIT** [reply]
I'm not sure why the same can't be done with declarative.

At the end of the day use the mode that you prefer. The end result is the same from sqlalchemy point of view

**Reply by qu3vipon** [reply]
In most cases, you don't separate the entity if you use declarative mapping. Are you saying that it's possible not to make DB I/O in the test case without separating the entity? If you really think so, please show me code examples.

**Reply by CaselIT** [reply]
That depends on the test. If a test does not make DB IO when using imperative mapping I don't see why it should do them when using declarative mapping.

**Reply by ArcticXWolf** [reply]
This was/is news to me too that you can use the declarative mapping without DB I/O.

I think you (maybe? I dont want to put words in your mouth) misunderstood @qu3vipon. With an imperative mapping we can design a test that just does not do the mapping and thus uses the models as simple python classes and thus have no DB I/O. So we can create a test without any SQLAlchemy involvement. With declarative mapping this is much harder since you cannot (as far as I know) "disable" the mapping (and thus DB I/O) for a test.

Please correct me, if I am wrong

**Reply by CaselIT** [reply]
Well mapped classes are just classes. You can instanciate them as you please and pass them around at will, without needing to associate them to a session.

**Reply by ArcticXWolf** [reply]
TIL

Okay, sorry for the confusion 🙂

**Reply by Rey092** [reply]
> "My own custom mapper" -> https://github.com/NEONKID/fastapi-ddd-example/blob/b6652bf5d0a61f693ce0ffb947e1e8cb331bd35c/modules/author/infrastructure/persistence/mapper.py#L11, simple Data Mapper pattern implementation, maping domain to persistance entity and back.
> 

This approach requires more code and undermines SQLAlchemy's Unit of Work. Imperative mapping is much better than these convoluted solutions.
